### ISSUE TITLE
License compliance when distributing software using Alpine Linux as base image

### ISSUE THREAD

**Issue by jgoeres** [alpinelinux/docker-alpine#3]
Hi,

we are currently investigating ways to ship our software as Docker images. There are mainly two challenges:
- we want to give our customers a full overview of what is inside our Docker images
- we want to be fully license compliant and not run into a variety of legal pitfalls. In particular, we need to have the complete source code of all open-source software. 
Using Alpine Linux as base image would be our preferred solution, and the apk tool lists a mere 14 packages installed for which we would need the source and the build instructions (like the APKBUILD files).

Alas, when looking at the Dockerfile I found that the base image is built by simply unpacking a tar.gz file with the root file system into the image, but I couldn't find any info on how this tar.gz's contents were created and what aside from those 14 packages (plus the APK package manager itself, I assume) was put into it.
Is there any info on that topic, or maybe a build script that creates the tar.gz file?

Regards
J

**Comment by jgoeres** [comment]
Meanwhile, we dug through the sources of the aports project (in particular the scripts directory) and pieced together the info on how the minirootfs tar.gz is build. I hope we are right to assume that these actually ARE the relevant scripts... can someone confirm that?

**Comment by ncopa** [comment]
Correct. the scripts directory in aports git repository is where the minirootfs is generated from.

**Comment by jgoeres** [comment]
THX for confirming this.
Is there any form of high-level documentation for the whole rootfs build process? We went through the sources and pieced the things together that end up in the minirootfs, but we could not go arbitrarily deep here.
This is the high-level summary of our research (we checked this for Alpine 3.7, but the scripts seem to change rarely, so this should be the same for more recent versions).

The main script is "mkimage.sh". It iterates over all scripts that follow the naming scheme mkimg.*.sh. The part matching the asterisk ("*") is called the "profile" - so let's call the mkimg.*.sh scripts the "profile scripts". Since we are only interested in the minirootfs profile, we focussed on "mkimg.minirootfs.sh".
Each "profile script" set various env vars, in particular the list of APKs to add. In case of the minirootfs the list is: busybox, alpine-baselayout, alpine-keys, apk-tools, libc-utils
Inside each of the "profile" scripts, mkimage.sh will call the "create_image_{OUTPUT_FORMAT} function, in the case of mkimg.minirootfs.sh this is "create_image_rootfs".
This function will run the genrootfs.sh script inside a fakeroot environment. What fakeroot basically does is make the command launched inside it see a subfolder of the "surrounding" linux system (let's call it the "build environment") and the software running in that fakeroot env will see this directory as if it were the root of the file system. 
What genrootfs.sh mainly does is use the APK tool of the build environment(!) to install the APK packages for the respective profile (plus their transitive dependencies) into a temporary location inside the fakeroot env(?).
Since the apk-tools package is the package providing the apk binary, this is also how APK itself is installed into the rootfs.
In addition to installing these packages, the genrootfs script adds a few other things to the minirootfs. These files are NOT downloaded from yet another place, but are generated by the script itself:
One section creates symbolic links for all "standard linux tools" that inside the Alpine world are handled by busybox (instead of the tool).
Further, a few configuration files are taken from the alpine-base package. Apparently this package is not "installed" with "apk add" but only fetched and unpacked to /etc, which is why alpine-base is NOT shown in the list of packages returned by "apk list" when inside the base image.

Then, some of the config (/etc/os-release, /etc/shadow) are modified to secure the image by preventing root logins, and by setting the actual Alpine version we are building here into the os-release file.
Then a repositories file is created, that tells apk inside the image from which repo it should get additional packages that people want to add in their own images.
Finally, the contents of the fakeroot directory are packaged into the resulting tar.gz file used in the Dockerfile.

Did we miss any essential parts that are relevant for the content of the resulting minirootfs?
